### PR TITLE
Fix #123 - Union types without sub-classes should be sealed

### LINKED ISSUE #123: Should FSharpList`1 be sealed?
originally opened at [codeplex](https://visualfsharp.codeplex.com/workitem/119) by [marten_range](https://www.codeplex.com/site/users/view/marten_range)

Inheriting a List<'T> in F# is forbidden:
```fsharp
type MyList() = 
    inherit List<int>()
// Compiler: error FS0945: Cannot inherit a sealed type
```
However creative C# developers can do this:
```csharp
class MyList : FSharpList<int> 
{
    MyList () : base(1, FSharpList<int>.Empty)
    {
    }

    public IEnumerator<int> GetEnumerator ()
    {
        yield break;
    }
}
````
IMHO FSharpList should include the sealed flag to prevent misuse. 

Adding sealed to FSharpList could, I guess, introduce a regression for developers that are overriding FSharpList but I suspect it's a quite rare regression. 

Mårten

>Comments
>dsyme wrote Aug 6, 2014 at 7:25 AM [x] 
>I would be happy for the "Sealed" flag to be added to this type.

>>marten_range wrote Aug 7, 2014 at 12:59 PM [x] 
>>Tried to add the [<Sealed>] attribute. Didn't work so I will need to dig more into how the assembly metadata is generated. Should be a good exercise.

>>marten_range wrote Aug 7, 2014 at 2:44 PM [x] 

>>I think I found a place where sealed can be injected for union cases. Seems to work for types like:
>>```fsharp
type X = | Y |  Z
```
>>However FSharpList1 despise looking like a union to me is generated differently so FSharpList1 don't get tagged sealed.

>>FSharpLìst1 special cases (like TailOrNull and HeadOrDefault) has surprised me before so if anyone has any tips on how to control the code generation for FSharpList1 that would be very interesting to me.

>>Thanks.

### PR DESCRIPTION
**Pull request by dsyme**
This implements [the F# Language User Voice feature request](http://fslang.uservoice.com/forums/245727-f-language/suggestions/7017131-make-f-union-types-like-list-that-have-no-subty) to change the F# compiled representations of union types by making union types sealed when their compiled representations have no sub-types.

This is also covered by suggestion #123.

Since this changes compiled forms in FSharp.Core.dll it is a candidate to consider for F# 4.0.

### CODE REVIEW

**Comment by latkin** [comment]
:+1:

File: tests/fsharp/core/array/test.fsx [hunk 1]
```diff
@@ -1368,6 +1368,28 @@ module bug872632 =
...
+module CheckUnionTypesAreSealed =
+    open System
+    do check "vwllfewlkefw1" (typedefof<list<int>>.IsSealed) true
+    do check "vwllfewlkefw2" (typedefof<option<int>>.IsSealed) true
+    type X1 = A | B
+    do check "vwllfewlkefw3" (typedefof<X1>.IsSealed) true
+    type X2 = A | B of string
+    do check "vwllfewlkefw4" (typedefof<X2>.IsSealed) false
+    type X3 = A | B | C
+    do check "vwllfewlkefw5" (typedefof<X3>.IsSealed) false
```
Comment (latkin) [inline tests/fsharp/core/array/test.fsx]: This returns `true` -- is it expected value or behavior that is incorrect?

File: tests/fsharp/core/array/test.fsx [hunk 2]
```diff
@@ -1368,6 +1368,28 @@ module bug872632 =
 
     do check "bug872632" Foo.x.Length 8
 
+module CheckUnionTypesAreSealed =
+    open System
+    do check "vwllfewlkefw1" (typedefof<list<int>>.IsSealed) true
```
Comment (latkin) [inline tests/fsharp/core/array/test.fsx]: `IsSealed` is not available in profiles 7, 78, or 259, which causes the tests in tests\fsharp\core\netcore to fail here

Comment (latkin) [inline tests/fsharp/core/array/test.fsx]: Quick fix
```fsharp
open System.Reflection

let isSealed (ty : Type) =
#if NetCore
    ty.GetTypeInfo().IsSealed
#else
    ty.IsSealed
#endif
```

**Comment by latkin** [comment]
Can you please update the codegen tests which are now failing? There are a bunch that use DUs which are now out of date.

**Comment by dsyme** [comment]
@latkin - Ok, will do 

I found it a bit tricky to update the one micro codegen test I added, because the *.il generated files get deleted by run RunTests.cmd process somewhere.  Is there an option to keep these around?  That would make it much simpler.

**Comment by latkin** [comment]
@dsyme yes this is a pain point. Let me see if there is something simple we could add to make that easier...

**Comment by dsyme** [comment]
baselines now updated

**Comment by latkin** [comment]
These are still failing for me:

- CodeGen\EmittedIL\SerializableAttribute (ToplevelNamespace.fs - Desktop)
- CodeGen\EmittedIL\SerializableAttribute (ToplevelNamespace.fs - Portable)
- Optimizations\GenericComparison (Compare05.fs - NetFx40)
- Optimizations\GenericComparison (Compare07.fs - NetFx40)
- Optimizations\GenericComparison (Compare10.fs - NetFx40)
- Optimizations\GenericComparison (Equals04.fs - NetFx40)
- Optimizations\GenericComparison (Equals06.fs - NetFx40)
- Optimizations\GenericComparison (Equals09.fs - NetFx40)
- Optimizations\GenericComparison (Hash05.fs - NetFx40)
- Optimizations\GenericComparison (Hash06.fs - NetFx40)
- Optimizations\GenericComparison (Hash09.fs - NetFx40)
- Optimizations\GenericComparison (Hash12.fs - NetFx40)

Detective work indicates the TopLevelNamespace guys were patched up by hand based on similar fix to TopLevelModule. However namespace guys turn out to need fix in more locations.

And I assume it was not realized that there are codegen tests under 'Optimizations'.

**Comment by dsyme** [comment]
Updated, thanks.  BTW if you get a chance could you look at the Perl problem in AppVeyor? https://ci.appveyor.com/project/KevinRansom/visualfsharp-radou/build/0.0.1.18

**Comment by dsyme** [comment]
Those "Optimization" codegen tests are still failing, will look at it in the morning :)

**Comment by latkin** [comment]
Don't forget the portable stuff in the core\array tests, too ;-)

**Comment by dsyme** [comment]
Now updated

**Pull request closed without merging**